File: app/src/main/java/gc/david/dfm/elevation/ElevationRemoteDataSource.java
package gc.david.dfm.elevation;

import com.google.gson.Gson;

import java.io.IOException;
import java.util.Locale;

import gc.david.dfm.elevation.model.ElevationModel;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

/**
 * Created by david on 05.01.17.
 */
public class ElevationRemoteDataSource implements ElevationRepository {

    private static final int ELEVATION_SAMPLES = 100;

    private final OkHttpClient client = new OkHttpClient();
    private final Gson         gson   = new Gson();

    @Override
    public ElevationModel getElevation(final String coordinatesPath) {
        final Request request = new Request.Builder().url(String.format(Locale.getDefault(),
                                                                        "https://maps.googleapis.com/maps/api/elevation/json?path=%s&samples=%d",
                                                                        coordinatesPath,
                                                                        ELEVATION_SAMPLES))
                                                     .header("content-type", "application/json")
                                                     .build();
        try {
            final Response response = client.newCall(request).execute();
            final ElevationModel elevationModel = gson.fromJson(response.body().charStream(), ElevationModel.class);
            return elevationModel;
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: elevation/src/main/java/gc/david/dfm/elevation/data/ElevationRemoteDataSource.kt
/*
 * Copyright (c) 2026 David Aguiar Gonzalez
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package gc.david.dfm.elevation.data

import com.google.gson.Gson
import gc.david.dfm.GeocodeApiKeyProvider
import gc.david.dfm.elevation.data.model.ElevationEntity
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.coroutines.executeAsync
import timber.log.Timber

class ElevationRemoteDataSource(geocodeApiKeyProvider: GeocodeApiKeyProvider) {

    private val client = OkHttpClient()
    private val gson = Gson()
    private val geocodeApiKey = geocodeApiKeyProvider.getApiKey()

    suspend fun getElevation(coordinatesPath: String, maxSamples: Int): ElevationEntity {
        val urlNoKey = "https://maps.googleapis.com/maps/api/elevation/json?path=$coordinatesPath&samples=$maxSamples"
        Timber.tag(TAG).d(urlNoKey)

        val url = "$urlNoKey&key=$geocodeApiKey"

        val request = Request.Builder().url(url)
                .header("content-type", "application/json")
                .build()
        val response = client.newCall(request).executeAsync()

        val elevationEntity =
                gson.fromJson(response.body.charStream(), ElevationEntity::class.java)
        return elevationEntity
    }

    companion object {

        private const val TAG = "ElevationRemoteDataSrc"
    }
}